source/simulator/core.cc

```diff
@@ -448,12 +448,16 @@ namespace aspect
 
     if (postprocess_manager.template has_matching_postprocessor<Postprocess::Particles<dim>>())
       {
-        particle_world = std::make_unique<Particle::World<dim>>();
-        if (SimulatorAccess<dim> *sim = dynamic_cast<SimulatorAccess<dim>*>(particle_world.get()))
-          sim->initialize_simulator (*this);
+        particle_worlds.emplace_back(std::move(std::make_unique<Particle::World<dim>>()));
+        for (unsigned int particle_world_index = 0 ; particle_world_index < particle_worlds.size(); ++particle_world_index)
```

Comment:
So this code makes sure that every element of the `particle_worlds` vector contains a valid World. Therefore in the code of the rest of this PR you do not need to check the entries of particle_worlds against `nullptr`.